micrometer-core/src/main/java/io/micrometer/core/instrument/binder/db/JooqMetrics.java

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright 2019 Pivotal Software, Inc.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.core.instrument.binder.db;
+
+import io.micrometer.core.annotation.Incubating;
+import io.micrometer.core.instrument.*;
+import io.micrometer.core.instrument.util.StringUtils;
+import org.jooq.*;
+import org.jooq.exception.DataAccessException;
+import org.jooq.impl.DSL;
+import org.jooq.impl.DefaultExecuteListener;
+
+/**
+ * Time SQL queries passing through JOOQ.
+ *
+ * @author Jon Schneider
+ * @since 1.4.0
+ */
+@Incubating(since = "1.4.0")
+public class JooqMetrics extends DefaultExecuteListener {
+  private static final String CONFIGURATION_KEY = "__micrometer_tag";
+  private static final String TIMER_SAMPLE_KEY = "__micrometer_timer_sample";
+
+  private final MeterRegistry registry;
+  private final Iterable<Tag> tags;
+
+  /**
+   * @param registry Registry to accumulate timings to.
+   */
+  public JooqMetrics(MeterRegistry registry) {
+    this(registry, Tags.empty());
+  }
+
+  /**
+   * @param registry Registry to accumulate timings to.
+   * @param tags     Tags to apply to all recorded metrics.
+   */
+  public JooqMetrics(MeterRegistry registry, Iterable<Tag> tags) {
+    this.registry = registry;
+    this.tags = tags;
+  }
+
+  @Override
+  public void start(ExecuteContext ctx) {
+    if (ctx.configuration().data(CONFIGURATION_KEY) != null) {
+      ctx.query().attach(ctx.configuration().derive());
+      ctx.query().configuration().data(TIMER_SAMPLE_KEY, Timer.start(registry));
```

Comment:
Instead of using the `Query.configuration().data()` for your context information, you could just use the `ExecuteContext.data()` map, instead. Benefits:

- It is guaranteed to be local to your single query execution
- You don't modify a query that might be produced from a cache in client code